mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt

```diff
@@ -812,38 +828,74 @@ class WidgetAdapter(
     }
 
     class ButtongridViewHolder internal constructor(private val initData: ViewHolderInitData) :
-        LabeledItemBaseViewHolder(initData, R.layout.widgetlist_buttongriditem), View.OnClickListener {
+        LabeledItemBaseViewHolder(initData, R.layout.widgetlist_buttongriditem),
+        View.OnClickListener,
+        View.OnTouchListener {
         private val table: GridLayout = itemView.findViewById(R.id.widget_content)
         private val spareViews = mutableListOf<MaterialButton>()
         private val maxColumns = itemView.resources.getInteger(R.integer.section_switch_max_buttons)
+        private var waitingForTouchRelease = false
+        private val buttonViews = mutableMapOf<String, MaterialButton>()
+
+        fun updateButton(button: Widget) {
+            buttonViews[button.id]?.let { buttonView ->
+                buttonView.setTextAndIcon(
+                    connection = connection, 
+                    label = button.label, 
+                    icon = button.icon,
+                    labelColor = button.labelColor,
+                    iconColor = button.iconColor,
+                    mapper = colorMapper
+                )
+                buttonView.isChecked = button.item?.state?.asString == button.command
+                boundWidget?.buttons?.replaceAll { if (it.id == button.id) button else it }
```

Comment:
This should be done in the SSE event listener, like for other widgets.

Reply:
> Sorry I still don't know my way around

Not a problem at all; the comments partially were meant as 'note to self' ;-)
I'm referring to [this code](https://github.com/openhab/openhab-android/blob/main/mobile/src/main/java/org/openhab/habdroid/ui/activity/PageConnectionHolderFragment.kt#L430)

Reply:
it now no longer keeps a separate copy of the Button widgets